File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
// class Solution {
//     public int maxProfit(int[] prices) {
//         int max=0;
//         for(int i=0;i<prices.length;i++){
//             for(int j=i+1;j<prices.length;j++){
//                 if(prices[j]>prices[i]){  
//                    int diff=prices[j]-prices[i];
//                     if(diff>max)
//                         max=diff;
//                 }
//             }
//         }
//         return max;
//     }
// }



class Solution {
    public int maxProfit(int[] prices) {
        int maxprofit=0;
        int sofar=prices[0];
        for(int i=0;i<prices.length;i++){
            sofar=Math.min(prices[i],sofar);
            int profit=prices[i]-sofar;
            maxprofit=Math.max(maxprofit,profit);
        }
        return maxprofit;
    }
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let left=new Array(prices.length);
    let right=new Array(prices.length);
    left[0]=prices[0];
    right[prices.length-1]=prices[prices.length-1];
    for(let i=1;i<prices.length;i++){
        left[i]=Math.min(left[i-1],prices[i]);
    }
    for(let i=prices.length-2;i>=0;i--){
        right[i]=Math.max(right[i+1],prices[i]);
    }
    let max=0;
    for(let i=0;i<prices.length;i++){
        max=Math.max(max,(right[i]-left[i]));
    }
    return max;
};
